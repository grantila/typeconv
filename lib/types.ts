
export interface Options
{
	declaration?: boolean;
	userPackage?: string;
	userPackageUrl?: string;
	noDisableLintHeader?: boolean;
	noDescriptiveHeader?: boolean;
}

export type SyncOrAsync< T > = T | PromiseLike< T >;

export type TypeImplementation =
	// TypeScript
	| 'ts'
	// JSON Schema
	| 'jsc'
	// GraphQL
	| 'gql'
	// Open API
	| 'oapi'
	// core-types
	| 'ct';
