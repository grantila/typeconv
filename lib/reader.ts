import type { NodeDocument, WarnFunction, ConversionResult } from "core-types"

import type { SyncOrAsync, TypeImplementation } from "./types.js"


export interface ReaderOptions
{
	warn: WarnFunction;
	filename?: string;
}

export type ReaderFunction = ( data: string, opts: ReaderOptions ) =>
	SyncOrAsync< ConversionResult< NodeDocument > >;

export type ShortcutReaderFunction = ( data: string, opts: ReaderOptions ) =>
	SyncOrAsync< ConversionResult< string > >;

export interface Reader
{
	kind: TypeImplementation;
	read: ReaderFunction;
	managedRead?: boolean;
	shortcut?: Partial< Record< TypeImplementation, ShortcutReaderFunction > >;
}
