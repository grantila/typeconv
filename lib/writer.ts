import type { NodeDocument, WarnFunction, ConversionResult } from "core-types"

import type { SyncOrAsync, TypeImplementation } from "./types.js"
import type { Reader } from "./reader.js"


export interface WriterOptions
{
	warn: WarnFunction;
	sourceFilename?: string;
	filename?: string;
	rawInput?: string;
}

export type WriterFunction =
	( doc: NodeDocument, opts: WriterOptions ) =>
		SyncOrAsync< ConversionResult< string > >;

export type Shortcut =
	(
		data: string,
		writeOpts: WriterOptions,
		reader: Reader
	) =>
		SyncOrAsync< ConversionResult< string > >;

export interface Writer
{
	kind: TypeImplementation;
	write: WriterFunction;
	shortcut?: Partial< Record< TypeImplementation, Shortcut > >;
}
