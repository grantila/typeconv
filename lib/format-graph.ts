import { TypeImplementation } from "./types.js"
import { Reader } from "./reader.js"
import { Writer } from "./writer.js"


type TI = TypeImplementation;

interface GraphPathSegment
{
	format: TI;
	reader: Reader;
	writer: Writer;
}

type PathKey = string;
type GraphPath = Array< GraphPathSegment >;

export function makePathKey( path: GraphPath ): PathKey
{
	return path
		.map( ( { format, reader, writer } ) =>
			`${reader.kind}->{${format}}->${writer.kind}`
		)
		.join( '  ' );
}

export class FormatGraph
{
	// Maps from-type -> (type-type -> reader)
	private readerGraph = new Map< TI, Map< TI, Reader > >( );
	// Maps from-type -> (type-type -> writer)
	private writerGraph = new Map< TI, Map< TI, Writer > >( );

	constructor( ) { }

	registerReader( reader: Reader )
	{
		const toMap: Array< [ TI, Reader ] > = [
			[ 'ct', reader ],
			...Object
				.keys( reader.shortcut ?? { } )
				.map( ( key ): [ TI, Reader ] => [ key as TI, reader ] ),
		];
		this.readerGraph.set( reader.kind, new Map( toMap ) );
	}

	registerWriter( writer: Writer )
	{
		const makeTo = ( ): [ TI, Writer ] => [ writer.kind, writer ];

		const insertWriter = ( from: TI ) =>
		{
			const old = [ ...this.writerGraph.get( from ) ?? [ ] ];
			this.writerGraph.set( from, new Map( [ ...old, makeTo( ) ] ) );
		};

		insertWriter( 'ct' );
		Object
			.keys( writer.shortcut ?? { } )
			.forEach( key => insertWriter( key as TI ) );
	}

	findAllPaths(
		reader: Reader,
		writer: Writer,
		shortcuts: boolean | undefined
	)
	: Array< GraphPath >
	{
		const paths = new Map< PathKey, GraphPath >( );

		interface Opts
		{
			allowManaged?: boolean;
			cache: Set< Reader >;
		}

		const appendSet = < T >( set: Set< T >, val: T ) =>
			new Set( [ ...set, val ] );

		const recurse = ( reader: Reader, path: GraphPath, opts: Opts ) =>
		{
			const { allowManaged = false, cache } = opts;

			const handleFound = ( writer: Writer, format: TI ) =>
			{
				if ( reader.managedRead && !allowManaged )
					return;
				const newPath = [ ...path, { reader, writer, format } ];
				const pathKey = makePathKey( newPath );
				paths.set( pathKey, newPath );
			};

			const formats = [
				...shortcuts ? [ ] : [ 'ct' ],
				...shortcuts !== false
					? Object.keys( reader.shortcut ?? { } )
					: [ ]
			];
			for ( const format of formats as Array< TI > )
			{
				const writers = this.writerGraph.get( format );
				for ( const [ to, _writer ] of writers?.entries( ) ?? [ ] )
				{
					if ( writer.kind === _writer.kind )
						handleFound( writer, format );
					else if ( reader.kind === _writer.kind )
						continue;
					else
					{
						// Find readers and recurse
						const readers = this.readerGraph.get( to );
						for ( const _reader of readers?.values( ) ?? [ ] )
						{
							if ( _reader.managedRead && !allowManaged )
								continue;
							else if ( cache.has( _reader ) )
								continue; // Cyclic
							recurse(
								_reader,
								[ ...path, { reader, writer: _writer, format } ],
								{
									cache: appendSet( cache, _reader ),
								}
							);
						}
					}
				}
			}
		};

		recurse( reader, [ ], { allowManaged: true, cache: new Set( ) } );

		return [ ...paths.values( ) ].sort( ( a, b ) => a.length - b.length );
	}

	findBestPath(
		reader: Reader,
		writer: Writer,
		shortcut: boolean | undefined
	)
	{
		const paths = this.findAllPaths( reader, writer, shortcut );
		if ( paths.length > 0 )
			return paths[ 0 ];

		// Allow all, find shortest path
		return this.findAllPaths( reader, writer, undefined )[ 0 ];
	}

	clone( )
	{
		const clone =  new FormatGraph( );

		const readers = [ ...this.readerGraph.values( ) ]
			.flatMap( map => [ ...map.values( ) ] );
		const writers = [ ...this.writerGraph.values( ) ]
			.flatMap( map => [ ...map.values( ) ] );

		readers.forEach( reader => clone.registerReader( reader ) );
		writers.forEach( writer => clone.registerWriter( writer ) );

		return clone;
	}
}

const defaultGraph = new FormatGraph( );

export function registerReader( reader: Reader )
{
	defaultGraph.registerReader( reader );
}
export function registerWriter( writer: Writer )
{
	defaultGraph.registerWriter( writer );
}

export interface ConversionOptions
{
	shortcut?: boolean;
}

export class ConversionContext
{
	private shortcut: boolean | undefined;
	private graph: FormatGraph;

	constructor(
		private reader: Reader,
		private writer: Writer,
		opts: ConversionOptions = { }
	)
	{
		this.shortcut = opts.shortcut;
		this.graph = defaultGraph.clone( );
		this.graph.registerReader( reader );
		this.graph.registerWriter( writer );
	}

	getPath( )
	{
		return this.graph.findBestPath(
			this.reader,
			this.writer,
			this.shortcut
		);
	}

}
