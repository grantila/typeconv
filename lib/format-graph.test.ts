import { FormatGraph, makePathKey } from "./format-graph.js"
import { Reader } from "./reader.js"
import { Writer } from "./writer.js"
import { TypeImplementation } from "./types.js"


function fakeReader(
	kind: TypeImplementation,
	formats: Array< TypeImplementation > = [ ],
	managedRead = false
)
: Reader
{
	return {
		kind,
		read: null as any,
		managedRead,
		shortcut: Object.fromEntries(
			formats.map( format => [ format, null ] )
		),
	};
}

function fakeWriter(
	kind: TypeImplementation,
	formats: Array< TypeImplementation > = [ ]
)
: Writer
{
	return {
		kind,
		write: null as any,
		shortcut: Object.fromEntries(
			formats.map( format => [ format, null ] )
		),
	};
}

describe( "format-graph", ( ) =>
{
	const gqlRead = fakeReader( 'gql', [ 'jsc' ] );
	const jscRead = fakeReader( 'jsc', [ 'jsc', 'oapi' ] );
	const oapiRead = fakeReader( 'oapi', [ 'jsc', 'oapi' ] );
	const tsRead = fakeReader( 'ts' );
	const stRead = fakeReader( 'st', [ 'jsc' ], true );

	const gqlWrite = fakeWriter( 'gql' );
	const tsWrite = fakeWriter( 'ts' );
	const jscWrite = fakeWriter( 'jsc', [ 'jsc' ] );
	const oapiWrite = fakeWriter( 'oapi', [ 'jsc' ] );
	const stWrite = fakeWriter( 'st', [ 'jsc' ] );

	const makeGraph = ( ) =>
	{
		const graph = new FormatGraph( );

		graph.registerReader( gqlRead );
		graph.registerReader( jscRead );
		graph.registerReader( oapiRead );
		graph.registerReader( tsRead );
		graph.registerReader( stRead );

		graph.registerWriter( gqlWrite );
		graph.registerWriter( tsWrite );
		graph.registerWriter( jscWrite );
		graph.registerWriter( oapiWrite );
		graph.registerWriter( stWrite );

		return graph;
	};

	it( "findAllPaths (incl core-types)", ( ) =>
	{
		const graph = makeGraph( );

		const paths = graph.findAllPaths(
			tsRead,
			gqlWrite,
			undefined
		);

		expect( paths.map( path => makePathKey( path ) ) ).toMatchSnapshot( );
	} );

	it( "findAllPaths (only shortcuts when none exist)", ( ) =>
	{
		const graph = makeGraph( );

		const paths = graph.findAllPaths(
			tsRead,
			gqlWrite,
			true
		);

		expect( paths.map( path => makePathKey( path ) ) ).toMatchSnapshot( );
} );

	it( "findAllPaths (only shortcuts when exist)", ( ) =>
	{
		const graph = makeGraph( );

		const paths = graph.findAllPaths(
			stRead,
			oapiWrite,
			true
		);

		expect( paths.map( path => makePathKey( path ) ) ).toMatchSnapshot( );
	} );

	it( "findAllPaths (no shortcuts)", ( ) =>
	{
		const graph = makeGraph( );

		const paths = graph.findAllPaths(
			stRead,
			oapiWrite,
			false
		);

		expect( paths.map( path => makePathKey( path ) ) ).toMatchSnapshot( );
	} );
} );
