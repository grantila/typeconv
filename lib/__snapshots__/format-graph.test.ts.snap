// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format-graph findAllPaths (incl core-types) 1`] = `
[
  "ts->{ct}->gql",
  "ts->{ct}->jsc  jsc->{ct}->gql",
  "ts->{ct}->oapi  oapi->{ct}->gql",
  "ts->{ct}->jsc  jsc->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->jsc  jsc->{ct}->oapi  oapi->{ct}->gql",
  "ts->{ct}->jsc  jsc->{jsc}->oapi  oapi->{ct}->gql",
  "ts->{ct}->oapi  oapi->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->oapi  oapi->{ct}->jsc  jsc->{ct}->gql",
  "ts->{ct}->oapi  oapi->{jsc}->jsc  jsc->{ct}->gql",
  "ts->{ct}->st  st->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->st  st->{ct}->jsc  jsc->{ct}->gql",
  "ts->{ct}->st  st->{ct}->oapi  oapi->{ct}->gql",
  "ts->{ct}->st  st->{jsc}->jsc  jsc->{ct}->gql",
  "ts->{ct}->st  st->{jsc}->oapi  oapi->{ct}->gql",
  "ts->{ct}->jsc  jsc->{ct}->ts  ts->{ct}->oapi  oapi->{ct}->gql",
  "ts->{ct}->jsc  jsc->{ct}->oapi  oapi->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->jsc  jsc->{jsc}->oapi  oapi->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->oapi  oapi->{ct}->ts  ts->{ct}->jsc  jsc->{ct}->gql",
  "ts->{ct}->oapi  oapi->{ct}->jsc  jsc->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->oapi  oapi->{jsc}->jsc  jsc->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->st  st->{ct}->ts  ts->{ct}->jsc  jsc->{ct}->gql",
  "ts->{ct}->st  st->{ct}->ts  ts->{ct}->oapi  oapi->{ct}->gql",
  "ts->{ct}->st  st->{ct}->jsc  jsc->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->st  st->{ct}->jsc  jsc->{ct}->oapi  oapi->{ct}->gql",
  "ts->{ct}->st  st->{ct}->jsc  jsc->{jsc}->oapi  oapi->{ct}->gql",
  "ts->{ct}->st  st->{ct}->oapi  oapi->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->st  st->{ct}->oapi  oapi->{ct}->jsc  jsc->{ct}->gql",
  "ts->{ct}->st  st->{ct}->oapi  oapi->{jsc}->jsc  jsc->{ct}->gql",
  "ts->{ct}->st  st->{jsc}->jsc  jsc->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->st  st->{jsc}->jsc  jsc->{ct}->oapi  oapi->{ct}->gql",
  "ts->{ct}->st  st->{jsc}->jsc  jsc->{jsc}->oapi  oapi->{ct}->gql",
  "ts->{ct}->st  st->{jsc}->oapi  oapi->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->st  st->{jsc}->oapi  oapi->{ct}->jsc  jsc->{ct}->gql",
  "ts->{ct}->st  st->{jsc}->oapi  oapi->{jsc}->jsc  jsc->{ct}->gql",
  "ts->{ct}->st  st->{ct}->ts  ts->{ct}->jsc  jsc->{ct}->oapi  oapi->{ct}->gql",
  "ts->{ct}->st  st->{ct}->ts  ts->{ct}->jsc  jsc->{jsc}->oapi  oapi->{ct}->gql",
  "ts->{ct}->st  st->{ct}->ts  ts->{ct}->oapi  oapi->{ct}->jsc  jsc->{ct}->gql",
  "ts->{ct}->st  st->{ct}->ts  ts->{ct}->oapi  oapi->{jsc}->jsc  jsc->{ct}->gql",
  "ts->{ct}->st  st->{ct}->jsc  jsc->{ct}->ts  ts->{ct}->oapi  oapi->{ct}->gql",
  "ts->{ct}->st  st->{ct}->jsc  jsc->{ct}->oapi  oapi->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->st  st->{ct}->jsc  jsc->{jsc}->oapi  oapi->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->st  st->{ct}->oapi  oapi->{ct}->ts  ts->{ct}->jsc  jsc->{ct}->gql",
  "ts->{ct}->st  st->{ct}->oapi  oapi->{ct}->jsc  jsc->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->st  st->{ct}->oapi  oapi->{jsc}->jsc  jsc->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->st  st->{jsc}->jsc  jsc->{ct}->ts  ts->{ct}->oapi  oapi->{ct}->gql",
  "ts->{ct}->st  st->{jsc}->jsc  jsc->{ct}->oapi  oapi->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->st  st->{jsc}->jsc  jsc->{jsc}->oapi  oapi->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->st  st->{jsc}->oapi  oapi->{ct}->ts  ts->{ct}->jsc  jsc->{ct}->gql",
  "ts->{ct}->st  st->{jsc}->oapi  oapi->{ct}->jsc  jsc->{ct}->ts  ts->{ct}->gql",
  "ts->{ct}->st  st->{jsc}->oapi  oapi->{jsc}->jsc  jsc->{ct}->ts  ts->{ct}->gql",
]
`;

exports[`format-graph findAllPaths (no shortcuts) 1`] = `
[
  "st->{ct}->oapi",
  "st->{ct}->gql  gql->{ct}->oapi",
  "st->{ct}->ts  ts->{ct}->oapi",
  "st->{ct}->jsc  jsc->{ct}->oapi",
  "st->{ct}->gql  gql->{ct}->ts  ts->{ct}->oapi",
  "st->{ct}->gql  gql->{ct}->jsc  jsc->{ct}->oapi",
  "st->{ct}->ts  ts->{ct}->gql  gql->{ct}->oapi",
  "st->{ct}->ts  ts->{ct}->jsc  jsc->{ct}->oapi",
  "st->{ct}->jsc  jsc->{ct}->gql  gql->{ct}->oapi",
  "st->{ct}->jsc  jsc->{ct}->ts  ts->{ct}->oapi",
  "st->{ct}->gql  gql->{ct}->ts  ts->{ct}->jsc  jsc->{ct}->oapi",
  "st->{ct}->gql  gql->{ct}->jsc  jsc->{ct}->ts  ts->{ct}->oapi",
  "st->{ct}->ts  ts->{ct}->gql  gql->{ct}->jsc  jsc->{ct}->oapi",
  "st->{ct}->ts  ts->{ct}->jsc  jsc->{ct}->gql  gql->{ct}->oapi",
  "st->{ct}->jsc  jsc->{ct}->gql  gql->{ct}->ts  ts->{ct}->oapi",
  "st->{ct}->jsc  jsc->{ct}->ts  ts->{ct}->gql  gql->{ct}->oapi",
]
`;

exports[`format-graph findAllPaths (only shortcuts when exist) 1`] = `
[
  "st->{jsc}->oapi",
  "st->{jsc}->jsc  jsc->{jsc}->oapi",
]
`;

exports[`format-graph findAllPaths (only shortcuts when none exist) 1`] = `[]`;
